server/src/main/java/org/elasticsearch/action/admin/indices/diskusage/TransportAnalyzeIndexDiskUsageAction.java

```diff
@@ -218,15 +219,23 @@ protected GroupShardsIterator<ShardIterator> shards(
         AnalyzeIndexDiskUsageRequest request,
         String[] concreteIndices
     ) {
-        final GroupShardsIterator<ShardIterator> groups = clusterService.operationRouting()
-            .searchShards(clusterState, concreteIndices, null, null);
-        for (ShardIterator group : groups) {
-            // fails fast if any non-active groups
-            if (group.size() == 0) {
-                throw new NoShardAvailableActionException(group.shardId());
+        final List<ShardIterator> groups = new ArrayList<>();
+        for (String index : concreteIndices) {
+            var indexMetadata = clusterState.metadata().index(index);
+            if (indexMetadata == null) {
+                throw new IndexNotFoundException(index);
+            }
+            for (int shardId = 0; shardId < indexMetadata.getNumberOfShards(); shardId++) {
+                var shardIterator = clusterState.routingTable().shardRoutingTable(index, shardId).activeInitializingShardsRandomIt();
```

Comment:
Doesn't this slightly change which shards are used to route this operation to in stateful? If yes, would it make sense to not change that and only return differently if stateless is enabled?